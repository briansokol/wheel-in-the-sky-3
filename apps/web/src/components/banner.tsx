import { TinyColor } from '@ctrl/tinycolor';
import { Segment } from '@repo/shared/classes/segment';
import { WheelManager } from '@repo/shared/classes/wheel-manager';
import { RefObject, useEffect, useMemo } from 'react';

interface BannerProps {
    bannerRef: RefObject<HTMLCanvasElement | null>;
    wheelManager: WheelManager | undefined;
    winner: Segment | undefined;
}

const FONT_NAME = '"Quicksand", sans-serif';

/**
 * Banner component that displays the wheel title and winner in a styled canvas
 * with rounded corners and triangular indicators pointing inward
 */
export function Banner({ bannerRef, wheelManager, winner }: BannerProps) {
    const fgColor = useMemo(() => {
        return winner?.textColor ?? '#ffffff';
    }, [winner?.textColor]);

    const segmentColor = useMemo(() => {
        return winner?.backgroundColor ?? '#000000';
    }, [winner?.backgroundColor]);

    const bgColor = useMemo(() => {
        if (!winner) {
            return '#000000';
        }

        const color = new TinyColor(winner.backgroundColor);
        const brightness = color.getBrightness();
        const isLight = color.isLight();

        if (isLight) {
            if (brightness > 215) {
                return color.darken(20).toString();
            } else {
                return color.brighten(20).toString();
            }
        } else {
            if (brightness < 30) {
                return color.brighten(20).toString();
            } else {
                return color.darken(20).toString();
            }
        }
    }, [winner]);

    useEffect(() => {
        if (bannerRef.current) {
            const ctx = bannerRef.current.getContext('2d');
            if (ctx) {
                const canvas = bannerRef.current;
                const canvasWidth = canvas.width;
                const canvasHeight = canvas.height;
                const cornerRadius = 15; // Radius for rounded corners

                // Clear the canvas to start with transparency
                ctx.clearRect(0, 0, canvasWidth, canvasHeight);

                // Create a rounded rectangle path for the background
                ctx.beginPath();
                ctx.moveTo(cornerRadius, 0);
                ctx.lineTo(canvasWidth - cornerRadius, 0);
                ctx.arcTo(canvasWidth, 0, canvasWidth, cornerRadius, cornerRadius);
                ctx.lineTo(canvasWidth, canvasHeight - cornerRadius);
                ctx.arcTo(canvasWidth, canvasHeight, canvasWidth - cornerRadius, canvasHeight, cornerRadius);
                ctx.lineTo(cornerRadius, canvasHeight);
                ctx.arcTo(0, canvasHeight, 0, canvasHeight - cornerRadius, cornerRadius);
                ctx.lineTo(0, cornerRadius);
                ctx.arcTo(0, 0, cornerRadius, 0, cornerRadius);
                ctx.closePath();

                // Fill the rounded rectangle background
                ctx.fillStyle = bgColor;
                ctx.fill();

                // Draw isosceles triangles pointing toward center
                const triangleWidth = 350;
                const triangleHeight = 250;
                const verticalPosition = canvasHeight / 2 - triangleHeight / 2;

                // Left triangle (pointing right)
                ctx.beginPath();
                ctx.fillStyle = segmentColor;
                ctx.moveTo(0, verticalPosition);
                ctx.lineTo(0, verticalPosition + triangleHeight);
                ctx.lineTo(triangleWidth, verticalPosition + triangleHeight / 2);
                ctx.closePath();
                ctx.fill();

                // Right triangle (pointing left)
                ctx.beginPath();
                ctx.fillStyle = segmentColor;
                ctx.moveTo(canvasWidth, verticalPosition);
                ctx.lineTo(canvasWidth, verticalPosition + triangleHeight);
                ctx.lineTo(canvasWidth - triangleWidth, verticalPosition + triangleHeight / 2);
                ctx.closePath();
                ctx.fill();

                // Set text properties for text
                ctx.fillStyle = fgColor;
                ctx.textAlign = 'center';
                ctx.textBaseline = 'middle';

                // Draw "Winner!" text
                ctx.font = `28px ${FONT_NAME}`;
                ctx.fillText('Winner!', canvasWidth / 2, canvasHeight / 3);

                // Draw title text
                ctx.fillText(wheelManager?.config.fullTitle ?? '', canvasWidth / 2, (canvasHeight * 2) / 3);

                // Draw winner name
                ctx.font = `48px ${FONT_NAME}`;
                ctx.fillText(winner?.name ?? '', canvasWidth / 2, canvasHeight / 2);

                // Draw app name
                ctx.font = `16px ${FONT_NAME}`;
                ctx.fillText('wheelinthesky.app', canvasWidth / 2, canvasHeight - 25);
            }
        }
    }, [bannerRef, bgColor, fgColor, segmentColor, wheelManager, winner]);

    return (
        <canvas
            ref={bannerRef}
            width="800"
            height="300"
            style={{
                borderRadius: '15px',
            }}
        ></canvas>
    );
}
