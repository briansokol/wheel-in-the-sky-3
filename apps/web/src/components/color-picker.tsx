import { Button, Input } from '@heroui/react';
import type { FormColorNameValue } from '@repo/shared/types/config';
import { debounce } from 'es-toolkit';
import { type FC, useCallback, useMemo } from 'react';
import { HexColorPicker } from 'react-colorful';
import { useFormContext } from 'react-hook-form';

interface ColorPickerProps {
    singleColorName: FormColorNameValue;
    multiColorName?: FormColorNameValue;
    allowMultiColor: boolean;
}

export const ColorPicker: FC<ColorPickerProps> = ({ singleColorName, multiColorName = '', allowMultiColor }) => {
    const { setValue, watch } = useFormContext();

    const debouncedSetColor = useMemo(() => {
        return debounce((value: string) => setValue(singleColorName, value), 100);
    }, [singleColorName, setValue]);

    const color: string = watch(singleColorName);
    const watchColors = watch(multiColorName);
    const colors: string[] = useMemo(
        () => (multiColorName !== undefined && watchColors !== undefined ? JSON.parse(watchColors) : []),
        [multiColorName, watchColors]
    );

    const addColor = useCallback(() => {
        if (multiColorName !== undefined && !colors.includes(color)) {
            const newColors = [...colors, color];
            setValue(multiColorName, JSON.stringify(newColors));
        }
    }, [color, colors, multiColorName, setValue]);

    const removeColor = useCallback(
        (index: number) => () => {
            if (multiColorName !== undefined) {
                const newColors = [...colors];
                newColors.splice(index, 1);
                setValue(multiColorName, JSON.stringify(newColors));
            }
        },
        [colors, multiColorName, setValue]
    );

    return (
        <>
            <HexColorPicker color={color} onChange={debouncedSetColor} style={{ width: '100%' }} />
            <div className="my-3">
                <Input
                    className="mr-3 inline-block w-[120px]"
                    type="text"
                    value={color}
                    onChange={(e) => setValue(singleColorName, e.target.value)}
                    aria-label="Chosen Color"
                />
                {allowMultiColor && multiColorName && (
                    <Button className="mr-3" onPress={addColor} color="default">
                        Add Color
                    </Button>
                )}
                {colors && colors.length > 0 && 'Click a color to remove it'}
            </div>
            {allowMultiColor && multiColorName && (
                <div data-testid="picker-color-list" className="flex flex-wrap">
                    {colors.map((color, i) => (
                        <Button
                            key={color}
                            className="relative mb-3 mr-3"
                            style={{ background: color }}
                            onPress={removeColor(i)}
                        ></Button>
                    ))}
                </div>
            )}
        </>
    );
};
