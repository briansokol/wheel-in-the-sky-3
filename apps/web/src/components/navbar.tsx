import {
    Link,
    Navbar,
    NavbarBrand,
    NavbarContent,
    NavbarItem,
    NavbarMenu,
    NavbarMenuItem,
    NavbarMenuToggle,
} from '@heroui/react';
import { useCallback, useMemo, useState } from 'react';
import { useLocation, useParams } from 'react-router';
import { PageBaseRoute } from '@/constants/routes';
import { isPage } from '@/utils/routes';

export function AppNavBar() {
    const { pathname } = useLocation();
    const params = useParams();
    const configId = params?.id ?? 'new';
    const uriEncodedConfigId = useMemo(() => encodeURIComponent(configId), [configId]);

    const [isMenuOpen, setIsMenuOpen] = useState(false);

    const closeMenu = useCallback(() => {
        setIsMenuOpen(false);
    }, [setIsMenuOpen]);

    const buildNavbarItems = useCallback(
        (Component: typeof NavbarItem | typeof NavbarMenuItem) => {
            return (
                <>
                    {configId !== 'new' && (
                        <Component isActive={isPage(pathname, PageBaseRoute.WheelV3)}>
                            <Link
                                color="foreground"
                                aria-current={isPage(pathname, PageBaseRoute.WheelV3) ? 'page' : 'false'}
                                href={`/wheel/v3/${uriEncodedConfigId}`}
                                onPress={closeMenu}
                            >
                                Wheel
                            </Link>
                        </Component>
                    )}
                    <Component isActive={isPage(pathname, PageBaseRoute.ConfigV3)}>
                        <Link
                            color="foreground"
                            aria-current={isPage(pathname, PageBaseRoute.ConfigV3) ? 'page' : 'false'}
                            href={`/config/v3/${uriEncodedConfigId}`}
                            onPress={closeMenu}
                        >
                            {configId === 'new' ? 'Create Wheel' : 'Configure'}
                        </Link>
                    </Component>
                    <Component isActive={isPage(pathname, PageBaseRoute.About)}>
                        <Link
                            color="foreground"
                            aria-current={isPage(pathname, PageBaseRoute.About) ? 'page' : 'false'}
                            href={`/about`}
                            onPress={closeMenu}
                        >
                            About
                        </Link>
                    </Component>
                </>
            );
        },
        [configId, uriEncodedConfigId, pathname, closeMenu]
    );

    const navbarItems = useMemo(() => {
        return buildNavbarItems(NavbarItem);
    }, [buildNavbarItems]);

    const navbarMenuItems = useMemo(() => {
        return buildNavbarItems(NavbarMenuItem);
    }, [buildNavbarItems]);

    return (
        <Navbar
            isMenuOpen={isMenuOpen}
            onMenuOpenChange={setIsMenuOpen}
            classNames={{
                item: [
                    'flex',
                    'relative',
                    'h-full',
                    'items-center',
                    "data-[active=true]:after:content-['']",
                    'data-[active=true]:after:absolute',
                    'data-[active=true]:after:bottom-0',
                    'data-[active=true]:after:left-0',
                    'data-[active=true]:after:right-0',
                    'data-[active=true]:after:h-[2px]',
                    'data-[active=true]:after:rounded-[2px]',
                    'data-[active=true]:after:bg-primary',
                ],
            }}
        >
            <NavbarMenuToggle aria-label={isMenuOpen ? 'Close Menu' : 'Open Menu'} className="sm:hidden" />
            <NavbarBrand>
                <Link
                    color="foreground"
                    aria-current={isPage(pathname, PageBaseRoute.Home) ? 'page' : 'false'}
                    href={`/`}
                >
                    <p className="font-bold text-inherit">Wheel in the Sky</p>
                </Link>
            </NavbarBrand>
            <NavbarContent className="hidden gap-4 sm:flex" justify="center">
                {navbarItems}
            </NavbarContent>
            <NavbarContent justify="end">
                {/*     <NavbarItem className="lg:flex">
                    <Dropdown>
                        <DropdownTrigger>
                            <Button variant="light" endContent={<FaAngleDown />}>
                                Favorites
                            </Button>
                        </DropdownTrigger>
                        <DropdownMenu aria-label="Favorite Wheel Actions" onAction={(key) => console.log(key)}>
                            <DropdownSection>
                                <DropdownItem key="add-favorite" color="success" endContent={<FaPlus />}>
                                    Add Favorite
                                </DropdownItem>
                                <DropdownItem
                                    key="remove-favorite"
                                    color="danger"
                                    endContent={<FaTrashCan />}
                                    showDivider
                                >
                                    Forget Favorite
                                </DropdownItem>
                            </DropdownSection>
                            <DropdownSection title="Favorite Wheels">
                                <DropdownItem
                                    key="wheel-1"
                                    description="Description of the wheel"
                                    endContent={<FaCheck />}
                                >
                                    Wheel 1
                                </DropdownItem>
                                <DropdownItem key="wheel-2" description="Description of the wheel">
                                    Wheel 2
                                </DropdownItem>
                            </DropdownSection>
                        </DropdownMenu>
                    </Dropdown>
                </NavbarItem> */}
            </NavbarContent>
            <NavbarMenu>{navbarMenuItems}</NavbarMenu>
        </Navbar>
    );
}
