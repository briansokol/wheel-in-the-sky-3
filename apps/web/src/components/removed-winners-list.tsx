import { Button } from '@heroui/react';
import { useCallback, useContext } from 'react';
import { MdRestore } from 'react-icons/md';
import { RemovedWinnersContext } from '@/contexts/removed-winners';

/**
 * Component that displays a list of removed winners with the ability to restore them.
 */
export function RemovedWinnersList() {
    const { removedWinners, setRemovedWinners } = useContext(RemovedWinnersContext);

    /**
     * Restores a winner by removing them from the removed winners list.
     *
     * @param {string} winner - The winner to restore
     * @returns {void}
     */
    const restoreWinner = useCallback(
        (winner: string) => {
            setRemovedWinners?.((prevWinners) => prevWinners.filter((w) => w !== winner));
        },
        [setRemovedWinners]
    );

    return (
        <div className="flex flex-col gap-2">
            {removedWinners.map((winner) => (
                <div key={winner} className="flex items-center justify-between">
                    <p className="max-w-[80%] truncate" title={winner}>
                        {winner}
                    </p>
                    <Button
                        isIconOnly
                        onPress={() => {
                            restoreWinner(winner);
                        }}
                    >
                        <MdRestore />
                    </Button>
                </div>
            ))}
            {removedWinners.length === 0 && (
                <div>
                    <p className="text-center text-gray-500">No removed winners</p>
                </div>
            )}
            <div>
                <Button
                    className="w-full"
                    isDisabled={removedWinners.length === 0}
                    startContent={<MdRestore />}
                    onPress={() => {
                        setRemovedWinners?.([]);
                    }}
                >
                    Restore All
                </Button>
            </div>
        </div>
    );
}
