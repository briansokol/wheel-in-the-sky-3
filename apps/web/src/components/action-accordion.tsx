import { Accordion, AccordionItem } from '@heroui/react';
import { PageBaseRoute } from '@/constants/routes';
import { isPage } from '@/utils/routes';
import { RemovedWinnersList } from './removed-winners-list';
import { SavedWheelsList } from './saved-wheels-list';

interface ActionAccordionProps {
    pathname: string;
    closeDrawer?: () => void;
}

export function ActionAccordion({ pathname, closeDrawer }: ActionAccordionProps) {
    return (
        <Accordion selectionMode="multiple" defaultExpandedKeys={['saved-wheels', 'removed-winners']}>
            <AccordionItem key="saved-wheels" aria-label="Saved Wheels" title="Saved Wheels">
                <SavedWheelsList closeDrawer={closeDrawer} />
            </AccordionItem>
            <AccordionItem key="removed-winners" aria-label="Removed Winners" title="Removed Winners">
                {isPage(pathname, PageBaseRoute.WheelV3) ? <RemovedWinnersList closeDrawer={closeDrawer} /> : null}
            </AccordionItem>
        </Accordion>
    );
}
