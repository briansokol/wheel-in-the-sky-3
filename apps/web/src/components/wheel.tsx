import { type WheelManager } from '@repo/shared/classes/wheel-manager';
import { type Coordinates, Easing, RotationDirection } from '@repo/shared/types/wheel';
import {
    type FC,
    type MouseEvent as ReactMouseEvent,
    type RefObject,
    useCallback,
    useContext,
    useEffect,
    useRef,
    useState,
} from 'react';
import { WheelSegment } from '@/components/wheel-segment';
import { RotationContext } from '@/contexts/rotation';
import { SegmentContext } from '@/contexts/segment';

interface WheelProps {
    wheelManager?: WheelManager;
    radius: string;
    isStatic?: boolean;
}

const MAX_SPEED = 30;
const MIN_SPEED = 0.1;
const MAX_BLUR = 10;

export const Wheel: FC<WheelProps> = ({ wheelManager, radius = '400px', isStatic = false }) => {
    const { rotation } = useContext(RotationContext);
    const { setHasWinner } = useContext(SegmentContext);

    const [rotationBlur, setRotationBlur] = useState(0);
    const [willChange, setWillChange] = useState(false);

    const mousePos = useRef<Coordinates>({ x: 0, y: 0 });
    const rotationDirection = useRef<RotationDirection>(RotationDirection.None);
    const startMousePos = useRef<Coordinates>({ x: 0, y: 0 });
    const prevMousePos = useRef<Coordinates>({ x: 0, y: 0 });
    const prevMouseAngle = useRef(0);
    const rotationDifference = useRef(0);
    const rotationSpeed = useRef(0);
    const easing = useRef(Easing.Out);
    const mouseDown = useRef(false);
    const quickClick = useRef(false);
    const frameId = useRef<number>(undefined);
    const wheelRef: RefObject<HTMLDivElement | null> = useRef(null);

    const getMouseAngle = useCallback(
        (mousePos: Coordinates): number => {
            const wheelPosition = wheelRef?.current?.getBoundingClientRect();
            if (!wheelPosition) {
                return 0;
            }

            const diffX = wheelPosition.left + wheelPosition.width / 2 - mousePos.x;
            const diffY = wheelPosition.top + wheelPosition.height / 2 - mousePos.y;

            return (Math.atan2(diffY, diffX) * 180) / Math.PI + 180;
        },
        [wheelRef]
    );

    const animate = useCallback(() => {
        if (wheelManager == undefined) {
            return;
        }
        if (rotationSpeed.current > 0.02) {
            wheelManager.setRotation((rotation) => {
                let newRotation = rotation + rotationSpeed.current * rotationDirection.current;
                if (newRotation < 0) {
                    newRotation += 360;
                } else if (newRotation >= 360) {
                    newRotation -= 360;
                }
                return newRotation;
            });
            rotationSpeed.current *= easing.current === Easing.In ? 1.01 : 0.99;

            setRotationBlur(
                rotationSpeed.current <= MIN_SPEED
                    ? 0
                    : rotationSpeed.current >= MAX_SPEED
                      ? MAX_BLUR
                      : ((rotationSpeed.current - MIN_SPEED) / (MAX_SPEED - MIN_SPEED)) * MAX_BLUR
            );
            frameId.current = requestAnimationFrame(animate);
        } else {
            setRotationBlur(0);
            frameId.current = undefined;
            rotationSpeed.current = 0;
            if (setHasWinner) {
                setHasWinner(true);
            }
            setWillChange(false);
        }
    }, [setHasWinner, wheelManager, setWillChange]);

    const handleMouseDown = useCallback(
        (e: ReactMouseEvent<HTMLDivElement, MouseEvent>) => {
            setWillChange(true);
            quickClick.current = true;
            setRotationBlur(0);
            setTimeout(() => {
                quickClick.current = false;
            }, 500);
            mouseDown.current = true;
            rotationDifference.current = rotation - getMouseAngle(mousePos.current);
            mousePos.current = { x: e.clientX, y: e.clientY };
            startMousePos.current = { x: e.clientX, y: e.clientY };
            if (frameId.current) {
                cancelAnimationFrame(frameId.current);
                frameId.current = undefined;
            }
            if (setHasWinner) {
                setHasWinner(false);
            }
        },
        [rotation, getMouseAngle, setHasWinner, setWillChange]
    );

    const handleMouseUp = useCallback(
        (e: MouseEvent) => {
            if (mouseDown.current) {
                mouseDown.current = false;
                if (e.clientX !== startMousePos.current.x || e.clientY !== startMousePos.current.y) {
                    easing.current = Easing.Out;
                    frameId.current = requestAnimationFrame(animate);
                } else if (quickClick.current) {
                    easing.current = Easing.In;
                    rotationSpeed.current = 0.8;
                    rotationDirection.current = RotationDirection.CounterClockwise;
                    frameId.current = requestAnimationFrame(animate);
                    setTimeout(() => {
                        rotationSpeed.current = 30 + Math.random() * 20;
                        rotationDirection.current = RotationDirection.Clockwise;
                        setTimeout(() => {
                            easing.current = Easing.Out;
                        }, 400);
                    }, 400);
                } else {
                    setRotationBlur(0);
                    setWillChange(false);
                    rotationSpeed.current = 0;
                    if (frameId.current) {
                        cancelAnimationFrame(frameId.current);
                        frameId.current = undefined;
                    }
                }
            }
        },
        [animate]
    );

    const handleMouseMove = useCallback(
        (e: MouseEvent) => {
            if (wheelManager == undefined) {
                return;
            }

            prevMousePos.current = mousePos.current;
            mousePos.current = { x: e.clientX, y: e.clientY };

            if (!mouseDown.current) {
                return;
            }

            if (setHasWinner) {
                setHasWinner(false);
            }

            const mouseAngle = getMouseAngle(mousePos.current);
            if (mouseAngle > prevMouseAngle.current || prevMouseAngle.current - mouseAngle > 180) {
                rotationDirection.current = RotationDirection.Clockwise;
            } else if (mouseAngle < prevMouseAngle.current || mouseAngle - prevMouseAngle.current > 180) {
                rotationDirection.current = RotationDirection.CounterClockwise;
            } else {
                rotationDirection.current = RotationDirection.None;
            }
            prevMouseAngle.current = mouseAngle;

            const speed = Math.hypot(
                mousePos.current.x - prevMousePos.current.x,
                mousePos.current.y - prevMousePos.current.y
            );
            rotationSpeed.current = speed * 0.5;

            let newRotation: number = mouseAngle + rotationDifference.current;
            if (newRotation < 0) {
                newRotation += 360;
            } else if (newRotation >= 360) {
                newRotation -= 360;
            }
            wheelManager.setRotation(newRotation);
        },
        [getMouseAngle, setHasWinner, wheelManager]
    );

    useEffect(() => {
        if (!isStatic) {
            globalThis.addEventListener('mousemove', handleMouseMove);

            return () => {
                globalThis.removeEventListener('mousemove', handleMouseMove);
            };
        }
    }, [isStatic, handleMouseMove]);

    useEffect(() => {
        if (!isStatic) {
            globalThis.addEventListener('mouseup', handleMouseUp);

            return () => {
                globalThis.removeEventListener('mouseup', handleMouseUp);
            };
        }
    }, [isStatic, handleMouseUp]);

    return (
        <div
            ref={wheelRef}
            onMouseDown={(e) => handleMouseDown(e)}
            className={`relative select-none overflow-hidden rounded-full ${isStatic ? '' : 'cursor-grab active:cursor-grabbing'}`}
            style={{
                width: radius,
                paddingTop: radius,
                clipPath: 'circle(50%)',
                transform: `rotate(${rotation ?? 0}deg)`,
                filter: `blur(0)`,
                willChange: willChange ? 'transform' : 'auto',
            }}
        >
            {wheelManager !== undefined &&
                wheelManager.segments.map((segment) => {
                    return <WheelSegment key={segment.id} segment={segment} />;
                })}
            <div
                className="absolute left-0 top-0 size-full"
                style={{
                    backdropFilter: `blur(${rotationBlur}px)`,
                    maskImage: 'radial-gradient(transparent 10%, black 50%)',
                    willChange: willChange ? 'backdropFilter' : 'auto',
                }}
            />
        </div>
    );
};
