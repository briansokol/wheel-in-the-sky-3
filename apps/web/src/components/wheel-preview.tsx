import { TinyColor } from '@ctrl/tinycolor';
import { Card } from '@heroui/react';
import { Config } from '@repo/shared/classes/config';
import { WheelManager } from '@repo/shared/classes/wheel-manager';
import { DEFAULT_FOREGROUND_COLOR, INVERSE_FOREGROUND_COLOR } from '@repo/shared/constants/colors';
import { PageColorType } from '@repo/shared/enums/page-colors';
import { WheelColorType } from '@repo/shared/enums/wheel-colors';
import { buildPageColorConfig, buildWheelColorConfig } from '@repo/shared/utils/configs';
import { FC, useContext, useEffect, useMemo } from 'react';
import { Wheel } from '@/components/wheel';
import { RotationContext } from '@/contexts/rotation';
import { useBgColor, useFgColor } from '@/hooks/colors';
import { calculateStartingRotation } from '@/utils/math';

interface WheelPreviewProps {
    title: string;
    description: string;
    names: string;
    randomizeOrder: boolean;
    wheelColorType: WheelColorType;
    baseColor?: string;
    colors?: string[];
    randomizeColor?: boolean;
    pageColorType: PageColorType;
    backgroundColor: string;
    showNames: boolean;
    isSelected?: boolean;
    height?: string;
}

export const WheelPreview: FC<WheelPreviewProps> = ({
    title,
    names,
    randomizeOrder,
    wheelColorType,
    baseColor,
    colors,
    randomizeColor,
    pageColorType,
    backgroundColor,
    showNames,
    height,
}) => {
    const { setRotation } = useContext(RotationContext);

    const [config, wheelManager] = useMemo(() => {
        const newConfig = new Config();
        newConfig.wheelColorConfig = buildWheelColorConfig(
            wheelColorType,
            baseColor ?? '',
            colors ?? [],
            randomizeColor ?? false
        );

        let foregroundColor = '';
        if (pageColorType === PageColorType.Single) {
            foregroundColor = new TinyColor(backgroundColor).isDark()
                ? DEFAULT_FOREGROUND_COLOR
                : INVERSE_FOREGROUND_COLOR;
        }

        newConfig.pageColorConfig = buildPageColorConfig(pageColorType, backgroundColor, foregroundColor);

        newConfig.setNames(names);
        newConfig.setRandomizeOrder(randomizeOrder);
        newConfig.setShowNames(showNames);

        const newWheelManager = new WheelManager();
        newWheelManager.init(newConfig, []);

        return [newConfig, newWheelManager];
    }, [
        names,
        randomizeOrder,
        wheelColorType,
        baseColor,
        colors,
        randomizeColor,
        pageColorType,
        backgroundColor,
        showNames,
    ]);

    useEffect(() => {
        if (setRotation && wheelManager) {
            setRotation(calculateStartingRotation(wheelManager?.segments?.length ?? 0));
        }
    }, [setRotation, wheelManager]);

    const bgColor = useBgColor(config, config.pageColorConfig.backgroundColor);
    const fgColor = useFgColor(config, config.pageColorConfig.foregroundColor);

    return (
        <Card className="w-full" shadow="lg" style={{ height: height ?? '400px', background: bgColor }}>
            {names.length > 0 ? (
                <div className="flex h-full flex-col items-center justify-center overflow-hidden">
                    <div
                        className="flex translate-y-8 items-center justify-center text-xl transition-all"
                        style={{ color: fgColor }}
                    >
                        {title}
                    </div>
                    <div className="flex size-full translate-y-28 items-center justify-center transition-all">
                        <Wheel wheelManager={wheelManager} radius="70%" isStatic={true} />
                    </div>
                </div>
            ) : (
                <div className="flex h-full items-center justify-center overflow-hidden" style={{ color: fgColor }}>
                    Add at Least 1 Wheel Segment
                </div>
            )}
        </Card>
    );
};
