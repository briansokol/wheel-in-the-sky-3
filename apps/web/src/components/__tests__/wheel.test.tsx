import { WheelManager } from '@repo/shared/classes/wheel-manager';
import { act, fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';
import { Wheel } from '@/components/wheel';
import { RotationProvider } from '@/contexts/rotation-provider';
import { SegmentProvider } from '@/contexts/segment-provider';

describe('Wheel', () => {
    const mockSetRotation = vi.fn();
    const mockWheelManager = {
        setRotation: mockSetRotation,
        segments: [
            {
                id: '1',
                name: 'Segment 1',
                backgroundColor: '#FF0000',
                textColor: '#FFFFFF',
                degrees: 180,
                index: 0,
            },
            {
                id: '2',
                name: 'Segment 2',
                backgroundColor: '#00FF00',
                textColor: '#000000',
                degrees: 180,
                index: 1,
            },
        ],
    } as unknown as WheelManager;

    beforeEach(() => {
        vi.useFakeTimers();
        mockSetRotation.mockClear();
    });

    afterEach(() => {
        vi.useRealTimers();
    });

    it('renders with correct initial properties', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        const wheelElement = container.firstChild as HTMLElement;
        expect(wheelElement).toHaveStyle({
            width: '400px',
            transform: 'rotate(0deg)',
        });
    });

    it('renders segments correctly', () => {
        const { getAllByText } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        expect(getAllByText(/Segment [12]/)).toHaveLength(2);
    });

    it('handles mouse down event', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        fireEvent.mouseDown(container.firstChild as HTMLElement, {
            clientX: 200,
            clientY: 200,
        });

        expect(mockSetRotation).not.toHaveBeenCalled();
    });

    it('handles mouse move event when mouse is down', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        fireEvent.mouseDown(container.firstChild as HTMLElement, {
            clientX: 200,
            clientY: 200,
        });

        fireEvent.mouseMove(document, {
            clientX: 220,
            clientY: 220,
        });

        expect(mockSetRotation).toHaveBeenCalled();
    });

    it('stops rotation animation on mouse down if wheel is spinning', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        const cancelAnimationFrameSpy = vi.spyOn(globalThis, 'cancelAnimationFrame');

        fireEvent.mouseDown(container.firstChild as HTMLElement, {
            clientX: 200,
            clientY: 200,
        });

        fireEvent.mouseMove(document, {
            clientX: 220,
            clientY: 220,
        });

        fireEvent.mouseUp(document);

        fireEvent.mouseDown(container.firstChild as HTMLElement, {
            clientX: 200,
            clientY: 200,
        });

        expect(cancelAnimationFrameSpy).toHaveBeenCalled();
    });

    it('does not stop rotation animation on mouse down if wheel is not spinning', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        const cancelAnimationFrameSpy = vi.spyOn(globalThis, 'cancelAnimationFrame');

        fireEvent.mouseDown(container.firstChild as HTMLElement, {
            clientX: 200,
            clientY: 200,
        });

        expect(cancelAnimationFrameSpy).not.toHaveBeenCalled();
    });

    it('does not attach event listeners when isStatic is true', () => {
        const addEventListenerSpy = vi.spyOn(globalThis, 'addEventListener');

        render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" isStatic={true} />
                </SegmentProvider>
            </RotationProvider>
        );

        expect(addEventListenerSpy).not.toHaveBeenCalled();
    });

    it('cleans up event listeners on unmount', () => {
        const removeEventListenerSpy = vi.spyOn(globalThis, 'removeEventListener');

        const { unmount } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        unmount();

        expect(removeEventListenerSpy).toHaveBeenCalledTimes(2); // mousedown, mousemove, mouseup
    });

    it('handles animation frame updates', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );

        fireEvent.mouseDown(container.firstChild as HTMLElement, {
            clientX: 200,
            clientY: 200,
        });

        fireEvent.mouseMove(document, {
            clientX: 220,
            clientY: 220,
        });

        fireEvent.mouseUp(document);

        act(() => {
            vi.runAllTimers();
        });

        expect(mockSetRotation).toHaveBeenCalled();
    });

    it('matches snapshot', () => {
        const { container } = render(
            <RotationProvider>
                <SegmentProvider>
                    <Wheel wheelManager={mockWheelManager} radius="400px" />
                </SegmentProvider>
            </RotationProvider>
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
