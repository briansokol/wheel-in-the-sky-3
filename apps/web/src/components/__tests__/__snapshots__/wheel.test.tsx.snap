// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wheel > matches snapshot 1`] = `
<div
  class="relative select-none overflow-hidden rounded-full cursor-grab active:cursor-grabbing"
  style="width: 400px; padding-top: 400px; clip-path: circle(50%); transform: rotate(0deg); filter: blur(0); will-change: auto;"
>
  <div
    class="absolute top-0 box-border size-full text-right text-black"
    data-testid="wheel-segment"
    style="background-color: rgb(255, 0, 0); transform: rotate(-90deg) rotate(90deg) rotate(0deg); transform-origin: 50% 50%; clip-path: polygon(100% 0, 100% 100%, 50% 100%, 50% 0);"
  >
    <div
      class="absolute right-5 top-1/2 text-right"
      data-testid="wheel-segment-name"
      style="color: rgb(255, 255, 255); transform-origin: right center; transform: translateY(-50%);"
    >
      Segment 1
    </div>
  </div>
  <div
    class="absolute top-0 box-border size-full text-right text-black"
    data-testid="wheel-segment"
    style="background-color: rgb(0, 255, 0); transform: rotate(-90deg) rotate(90deg) rotate(180deg); transform-origin: 50% 50%; clip-path: polygon(100% 0, 100% 100%, 50% 100%, 50% 0);"
  >
    <div
      class="absolute right-5 top-1/2 text-right"
      data-testid="wheel-segment-name"
      style="color: rgb(0, 0, 0); transform-origin: right center; transform: translateY(-50%);"
    >
      Segment 2
    </div>
  </div>
  <div
    class="absolute left-0 top-0 size-full"
    style="mask-image: radial-gradient(transparent 10%, black 50%); will-change: auto;"
  />
</div>
`;
